### PR TITLE
fix#3513

### PR DESCRIPTION
**Pull request by jlp765**
Added random number to make statement identifier unique
Added tags to improve effect tracking

### CODE REVIEW

File: lib/impure/db_postgres.nim [hunk 1]
```diff
@@ -168,8 +169,8 @@ proc newRow(L: int): Row =
 proc setupQuery(db: DbConn, query: SqlQuery,
                 args: varargs[string]): PPGresult =
   # s is a dummy unique id str for each setupQuery query
-  let s = "setupQuery_Query_" & string(query)
-  var res = pqprepare(db, s, dbFormat(query, args), 0, nil)
+  let s = "Q_" & $random(100000) & "_" & string(query)
+  discard pqprepare(db, s, dbFormat(query, args), 0, nil)
```
Comment (dom96) [inline lib/impure/db_postgres.nim]: Why not check that the return value isn't an error?

File: lib/impure/db_postgres.nim [hunk 2]
```diff
@@ -168,8 +169,8 @@ proc newRow(L: int): Row =
 proc setupQuery(db: DbConn, query: SqlQuery,
                 args: varargs[string]): PPGresult =
   # s is a dummy unique id str for each setupQuery query
-  let s = "setupQuery_Query_" & string(query)
-  var res = pqprepare(db, s, dbFormat(query, args), 0, nil)
+  let s = "Q_" & $random(100000) & "_" & string(query)
```
Comment (dom96) [inline lib/impure/db_postgres.nim]: Maybe generate an `Oid` to ensure randomness: http://nim-lang.org/docs/oids.html

**Comment by jlp765** [comment]
Thanks for that feedback.  Changes made.

**Comment by vegansk** [comment]
You can't use the prepared statement from this version of ``setupQuery`` twice. So, what is the reason to use prepared statement?

**Comment by jlp765** [comment]
From the Postgres Doco:

`The function creates a prepared statement named stmtName from the query string, which must contain a single SQL command. stmtName can be "" to create an unnamed statement, in which case any pre-existing unnamed statement is automatically replaced; otherwise it is an error if the statement name is already defined in the current session. 
`

So we could set the statement name to be an empty string.  I like the unique name via an Oid().

Also, there is nothing stopping you using the lower level ``postges`` wrapper library, to prepare a postgres statement and then call it repeatedly if needed.  
However, I doubt it will save you much in execution time, compared to the ease of using the ``db_postgres`` lib

**Comment by vegansk** [comment]
I can't agree with you. This solution will lead to the resource leaks. How can we use db connection pool in application server for example?

**Comment by dom96** [comment]
Perhaps we should create an unnamed statement, that would stop any resource leaks.

**Comment by vegansk** [comment]
Checked both. On my system (I7, SSD, 8GB, Linux x64), variant with unnamed prepared statements is 1.5 times slower than pqexec variant. Just simple select 10000 times:

Unnamed prepared statement:
```
I, [2015-11-11T16:04:47] -- query_test: Connected to postgresql
I, [2015-11-11T16:04:59] -- query_test: Closing the connection
        Command being timed: "./query_test"
        User time (seconds): 1.68
        System time (seconds): 2.25
        Percent of CPU this job got: 31%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:12.34
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 348552
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 87294
        Voluntary context switches: 198983
        Involuntary context switches: 1709
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

psexec:
```
I, [2015-11-11T16:07:24] -- query_test_no_ps: Connected to postgresql
I, [2015-11-11T16:07:32] -- query_test_no_ps: Closing the connection
        Command being timed: "./query_test_no_ps"
        User time (seconds): 1.13
        System time (seconds): 1.15
        Percent of CPU this job got: 29%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:07.80
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 328260
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 82221
        Voluntary context switches: 99613
        Involuntary context switches: 814
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

**Comment by Araq** [comment]
Can we get rid of the prepared statements please? What problem do they solve? There is no caching layer that could cache these queries properly, that's Postgre's business anyway...

**Comment by dom96** [comment]
Closing in favour of #3514.

**Pull request closed without merging**